File: scripts/generate-pi.ts
import { writeFileSync, mkdirSync, existsSync } from "fs";
import { join } from "path";

const NUM_DIGITS = 1_000_000;
const GUARD = 20;

/**
 * Compute arccot(x) = arctan(1/x) in fixed-point BigInt arithmetic.
 * Returns arctan(1/x) × 10^precision.
 */
function arccot(x: bigint, precision: number): bigint {
  const unity = 10n ** BigInt(precision);
  const negX2 = -(x * x); // sign flip gives alternating series
  let sum = 0n;
  let term = unity / x;
  let n = 1n;

  while (term !== 0n) {
    sum += term / n;
    term /= negX2;
    n += 2n;
  }
  return sum;
}

console.log(
  `Computing ${NUM_DIGITS.toLocaleString()} decimal digits of π (Machin's formula)...`
);
console.time("computation");

const precision = NUM_DIGITS + GUARD;

// Machin's formula: π/4 = 4·arctan(1/5) − arctan(1/239)
const pi = 4n * (4n * arccot(5n, precision) - arccot(239n, precision));

const piStr = pi.toString();
const decimals = piStr.substring(1, NUM_DIGITS + 1); // skip leading "3"

console.timeEnd("computation");
console.log(
  `Verify: 3.${decimals.substring(0, 20)} (expect 3.14159265358979323846)`
);

const outDir = join(process.cwd(), "public");
if (!existsSync(outDir)) mkdirSync(outDir, { recursive: true });

const outPath = join(outDir, "pi-digits.txt");
writeFileSync(outPath, decimals, "utf-8");
console.log(
  `Written ${decimals.length.toLocaleString()} digits to public/pi-digits.txt`
);

File: scripts/generate-pi.py
"""Generate 1 million decimal digits of pi using mpmath (Chudnovsky algorithm)."""

import time
import os
from mpmath import mp

NUM_DIGITS = 1_000_000

print(f"Computing {NUM_DIGITS:,} decimal digits of π (Chudnovsky via mpmath)...")
start = time.time()

mp.dps = NUM_DIGITS + 10  # extra guard digits
pi_str = mp.nstr(mp.pi, NUM_DIGITS + 1)  # "3.14159..."
decimals = pi_str[2:]  # strip "3."

elapsed = time.time() - start
print(f"Computation: {elapsed:.1f}s")
print(f"Verify: 3.{decimals[:20]} (expect 3.14159265358979323846)")

out_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "public")
os.makedirs(out_dir, exist_ok=True)

out_path = os.path.join(out_dir, "pi-digits.txt")
with open(out_path, "w") as f:
    f.write(decimals)

print(f"Written {len(decimals):,} digits to public/pi-digits.txt")
